### ISSUE TITLE
Kiali dashboard login - Unauthorized

### ISSUE THREAD

**Issue by rnkhouse** [istio/istio#17527]
**Bug description**
I created `kiali` secret and enabled it using this guideline https://istio.io/docs/tasks/telemetry/kiali/
But, when I execute `kubectl -n istio-system port-forward $(kubectl -n istio-system get pod -l app=kiali -o jsonpath='{.items[0].metadata.name}') 20001:20001` getting below error

	Forwarding from 127.0.0.1:20001 -> 20001
	Forwarding from [::1]:20001 -> 20001
	Handling connection for 20001
	E1001 11:06:08.763555   17720 portforward.go:303] error copying from remote stream to local connection: readfrom tcp6 [::1]:20001->[::1]:57632: write tcp6 [::1]:20001->[::1]:57632: wsasend: An established connection was aborted by the software in your host machine.
	Handling connection for 20001
	Handling connection for 20001
	Handling connection for 20001
	E1001 11:06:09.088557   17720 portforward.go:303] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:20001->127.0.0.1:57636: write tcp4 127.0.0.1:20001->127.0.0.1:57636: wsasend: An established connection was aborted by the software in your host machine.
	Handling connection for 20001
	E1001 11:06:12.078569   17720 portforward.go:303] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:20001->127.0.0.1:57638: write tcp4 127.0.0.1:20001->127.0.0.1:57638: wsasend: An established connection was aborted by the software in your host machine.
	E1001 11:06:39.386956   17720 portforward.go:293] error creating forwarding stream for port 20001 -> 20001: Timeout occured

Also, credential that I created is not working as well:

>Unauthorized. The provided credentials are not valid to access Kiali. Please check your credentials and try again.

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[x] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
I should be able to access `kiali` dashboard.

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
kubernetes version: v.1.12.4
istio version: v.1.2.6

**How was Istio installed?**
helm template

**Environment where bug was observed (cloud vendor, OS, etc)**
AKS

**Comment by israel-hdez** [comment]
@rnkhouse Question: are you getting the errors when you run the port-forward command? or are you getting the errors when you try to access Kiali in the browser?

**Comment by rnkhouse** [comment]
@israel-hdez I mentioned command line logs in questions and also on the browser I am getting `Unauthorized` error even though I am using the correct credentials.

**Comment by israel-hdez** [comment]
Hi, again @rnkhouse 
I've followed the install instructions using Istio 1.2.7 and I can successfully login.

The only thing that comes into my mind is that, probably, the secret was created with carriage returns in the username and password. Try these commands to check:

```bash
# Both commands must print 0 (zero)
kubectl get secret -n istio-system kiali -o jsonpath='{.data.passphrase}' | base64 -d | wc -l
kubectl get secret -n istio-system kiali -o jsonpath='{.data.username}' | base64 -d | wc -l
```

Anyway, I recommend you to use the latest kiali by installing with these instructions: https://www.kiali.io/documentation/getting-started/

**Comment by rnkhouse** [comment]
@israel-hdez Both command prints 0. But, still getting the same error on browser: `Unauthorized. The provided credentials are not valid to access Kiali. Please check your credentials and try again.` I am using `v1.2.6` of istio.

**Comment by israel-hdez** [comment]
@rnkhouse can you, please, paste here the logs of the kiali pod?

**Comment by rnkhouse** [comment]
@israel-hdez here it is.

	I1008 12:02:24.493725       1 kiali.go:67] Kiali: Version: v0.20.0, Commit: 140969bc2ef1a77ca7387b47f8b23dcbd4818bd0
	I1008 12:02:24.572089       1 kiali.go:68] DEBUG: Kiali: Command line: [/opt/kiali/kiali -config /kiali-configuration/config.yaml -v 4]
	I1008 12:02:24.593279       1 config.go:461] DEBUG: Reading YAML config from [/kiali-configuration/config.yaml]
	I1008 12:02:24.603318       1 kiali.go:203] Using authentication strategy [login]
	I1008 12:02:24.632669       1 kiali.go:88] Kiali: Console version: 0.20.0
	I1008 12:02:24.632686       1 kiali.go:263] Updating base URL in index.html with [/kiali]
	I1008 12:02:24.652975       1 kiali.go:276] DEBUG: Base URL has been updated to [<base href="/kiali/"]
	I1008 12:02:24.733990       1 kiali.go:244] Generating env.js from config
	I1008 12:02:24.734039       1 kiali.go:249] DEBUG: The content of /opt/kiali/console/env.js will be:
	window.WEB_ROOT='/kiali';
	E1008 12:02:24.782249       1 kiali.go:106] Jaeger is not available : Get http:///api/services: http: no Host in request URL
	I1008 12:02:24.812793       1 server.go:49] Server endpoint will start at [:20001/kiali]
	I1008 12:02:24.812808       1 server.go:50] Server endpoint will serve static content from [/opt/kiali/console]
	I1008 12:02:24.812821       1 metrics_server.go:18] Starting Metrics Server on [:9090]
	I1008 12:02:24.859297       1 kiali.go:135] Secret is now available.
	I1008 18:43:06.817930       1 authentication.go:101] DEBUG: (Re-)authentication was asked. Validation of old token failed with: token contains an invalid number of segments
	I1008 18:43:09.579353       1 authentication.go:101] DEBUG: (Re-)authentication was asked. Validation of old token failed with: token contains an invalid number of segments
	I1008 18:43:10.753877       1 authentication.go:101] DEBUG: (Re-)authentication was asked. Validation of old token failed with: token contains an invalid number of segments
	I1008 18:43:11.256472       1 authentication.go:101] DEBUG: (Re-)authentication was asked. Validation of old token failed with: token contains an invalid number of segments

**Comment by jmazzitelli** [comment]
what's the command you used to create the secret? Can you post the YAML for the secret?

**Comment by jmazzitelli** [comment]
>I1008 18:43:06.817930       1 authentication.go:101] DEBUG: (Re-)authentication was asked. Validation of old token failed with: token contains an invalid number of segments

Purge your browser of the Kiali cookies - looks like an old cookie might be cached. To see if this could be the issue, open up an incognito window and try to log into Kiali.

**Comment by rnkhouse** [comment]
@jmazzitelli 

	---
	apiVersion: v1
	kind: Secret
	metadata:
	  name: kiali
	  namespace: istio-system
	  labels:
	    app: kiali
	type: Opaque
	data:
	  username: <KIALI_USERNAME>
	  passphrase: <KIALI_PASSPHRASE>

**Comment by jmazzitelli** [comment]
>   username: <KIALI_USERNAME>
>   passphrase: <KIALI_PASSPHRASE>

Yeah, that's the important part - wanted to see what your hash codes were (are you using the "admin" / "admin" credentials?) I am wondering how you created it - did you ensure you did not include a newline when you encoded the username and passphrase?

**Comment by rnkhouse** [comment]
@jmazzitelli 
I am replacing that string with this script. And yes making sure that new line is not part of the string as well.

    istio_kiali_username_base64=$(echo ${istio_kiali_username} | tr -d '\n' | base64)
    istio_kiali_passphrase_base64=$(echo ${istio_kiali_passphrase} | tr -d '\n' | base64)
    sed -i "s|KIALI_USERNAME|${istio_kiali_username_base64}|g" istio_kiali_secret.yaml
    sed -i "s|KIALI_PASSPHRASE|${istio_kiali_passphrase_base64}|g" istio_kiali_secret.yaml

**Comment by jmazzitelli** [comment]
ok, well, I cannot vouch for that.. rather than assume that script is doing what you expect, can you use the example in the telemetry/kiali docs and see if that works. If it STILL fails, then we know it isn't the secret:

https://istio.io/docs/tasks/telemetry/kiali/#create-a-secret

or just use the demo secret here (which encodes admin/admin): 

https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/charts/kiali/templates/demosecret.yaml#L2-L15

**Comment by rnkhouse** [comment]
It worked when I deleted the kiali deployment and re-deploy it.

**Comment by sabiha969** [comment]
I have deleted kiali, how to redeploy it ?

**Comment by jmazzitelli** [comment]
> I have deleted kiali, how to redeploy it ?

https://www.kiali.io/documentation/getting-started

**Comment by sabiha969** [comment]
How to redeploy kiali deployment, in kubernetes. I mean to say.

**Comment by jmazzitelli** [comment]
> How to redeploy kiali deployment, in kubernetes. I mean to say.

Same link as before. Kiali supports Kubernetes deployment using the same instructions I linked earlier.

**Comment by wuestkamp** [comment]
If you change the secret, make sure to delete the kiali pod so a new one gets created and the new secret will be mounted.

**Comment by sghaida** [comment]
![image](https://user-images.githubusercontent.com/888275/73303378-b022b600-4216-11ea-8fce-630b3136a8a9.png)

still happens, version: 1.9.1 (kiali) istio: 1.4.3

![image](https://user-images.githubusercontent.com/888275/73303567-fed05000-4216-11ea-8850-11f7af13e194.png)

**Comment by sghaida** [comment]
kiali pod log:
2020-01-28T20:36:08.96110042Z I0128 20:36:08.958532       1 kiali.go:66] Kiali: Version: v1.9.1, Commit: 2431f9b8844202e27ec1c92f404fe7485253ffec
2020-01-28T20:36:08.96114112Z I0128 20:36:08.958779       1 kiali.go:210] Using authentication strategy [login]
2020-01-28T20:36:08.96114602Z I0128 20:36:08.958807       1 kiali.go:87] Kiali: Console version: 1.9.1
2020-01-28T20:36:08.96114922Z I0128 20:36:08.958813       1 kiali.go:272] Updating base URL in index.html with [/kiali]
2020-01-28T20:36:08.96115242Z I0128 20:36:08.959159       1 kiali.go:253] Generating env.js from config
2020-01-28T20:36:08.997048502Z I0128 20:36:08.996376       1 server.go:57] Server endpoint will start at [:20001/kiali]
2020-01-28T20:36:08.997071702Z I0128 20:36:08.996393       1 server.go:58] Server endpoint will serve static content from [/opt/kiali/console]
2020-01-28T20:36:08.997075302Z I0128 20:36:08.996405       1 metrics_server.go:18] Starting Metrics Server on [:9090]
2020-01-28T20:36:08.997078202Z I0128 20:36:08.996496       1 kiali.go:142] Secret is now available.

**Comment by israel-hdez** [comment]
@sghaida If you are creating the secret manually and using `echo` command, please make sure that you pass the `-n` flag to `echo`; e.g:

```bash
% echo -n 'admin' | base64
YWRtaW4=
```

Without the `-n` flag the encoded result is different:

```bash
% echo 'admin' | base64 
YWRtaW4K
```

If you still have the issue, please first upgrade to latest Kiali which is version 1.12. If latest Kiali still has the issue, please open an issue in Kiali repository: https://github.com/kiali/kiali/issues/new/choose

**Comment by fafg** [comment]
actually i got the same issue, but I wasn't using the echo, but I've pushed the echo to my command chain and it worked. thank you very much.

**Comment by huegelc** [comment]
Issue still happens in Rancher 2.3.5 with Istio 1.4.6 https://github.com/rancher/rancher/issues/25527

**Comment by israel-hdez** [comment]
@huegelc 

In the rancher issue, I see you provided this for configmap:

```yaml
auth:
strategy: login
```

It doesn't have the right indentation. It should be:

```yaml
auth:
  strategy: login
```

I'm not sure if this is really the issue, but it's worth to check.

**Comment by jmazzitelli** [comment]
I don't think that is his problem @israel-hdez - I think its just the way his comment was formatted. Look at his server logs - shows its using "login" strategy:

```
2020-01-28T20:36:08.96114112Z I0128 20:36:08.958779 1 kiali.go:210] Using authentication strategy [login]
```

It is possible the secret was created after the server was started, or the password is incorrect in the secret? I dunno - but we have not seen any issues (outside of user-error) regarding login strategy not logging people in when they provide correct credentials.

**Comment by israel-hdez** [comment]
OK @jmazzitelli, that makes sense. So, if indentation is not the issue, I see you @huegelc already saw that the secrets are mounted inside the pod in `/kiali-secret/`.
Perhaps you could `cat` those files to check that they have the right values?

**Comment by huegelc** [comment]
@israel-hdez yes, the indentation got messed up when copy&paste. Here are the actual objects:
```
kubectl get cm -n istio-system kiali -o yaml
apiVersion: v1
data:
  config.yaml: |-
    istio_namespace: istio-system
    deployment:
      accessible_namespaces: ['**']
    auth:
      strategy: login
    server:
      port: 20001
      web_root: /
    external_services:
      tracing:
        url: http://tracing.istio-system:80
      grafana:
        url: http://access-grafana.cattle-prometheus:80
      prometheus:
        url: http://access-prometheus.cattle-prometheus:80
kind: ConfigMap
metadata:
  creationTimestamp: "2020-02-22T08:39:35Z"
  labels:
    app: kiali
    chart: kiali
    heritage: Tiller
    io.cattle.field/appId: cluster-istio
    release: cluster-istio
  name: kiali
  namespace: istio-system
  resourceVersion: "14488034"
  selfLink: /api/v1/namespaces/istio-system/configmaps/kiali
  uid: 58f62a96-d980-43e4-b5d9-278fac6486f4
```

```
kubectl exec -it -n istio-system kiali-75db5d5d46-tptbt -- cat /kiali-secret/passphrase
admin

kubexec -it -n istio-system kiali-75db5d5d46-tptbt -- cat /kiali-secret/username
admin
```

```
kubectl get secrets -n istio-system kiali -o yaml
apiVersion: v1
data:
  passphrase: YWRtaW4=
  username: YWRtaW4=
kind: Secret
metadata:
  creationTimestamp: "2020-02-22T09:07:11Z"
  labels:
    app: kiali
  name: kiali
  namespace: istio-system
  resourceVersion: "19213974"
  selfLink: /api/v1/namespaces/istio-system/secrets/kiali
  uid: c424ec3d-2f4e-480a-a436-f0e68e4a7113
type: Opaque

```
```
kubectl get secrets -n istio-system kiali -o jsonpath='{.data.passphrase}' | base64 -d
admin

kubectl get secrets -n istio-system kiali -o jsonpath='{.data.username}' | base64 -d
admin
```
Thanks

**Comment by israel-hdez** [comment]
@huegelc Everything looks fine. You should be able to login using 'admin' for both username and password.

But this is confusing. These are the default configs, while in the rancher ticket the secret is customized. May be you were trying to login with the wrong password?

**Comment by israel-hdez** [comment]
@huegelc how are you exposing and accessing Kiali?

**Comment by huegelc** [comment]
@israel-hdez well, yes I've changed the passphrase as I thought there is something wrong with it. The current username/passphrase is like posted above (yes, I re-deployed the kiali container)

I'm accessing the Kiali dashboard from within Rancher:

![image](https://user-images.githubusercontent.com/6955465/76658613-c9888300-6574-11ea-900b-fa690aa1c13f.png)

**Comment by israel-hdez** [comment]
@huegelc I think you will need to wait on Rancher's support. They may have an additional security layer in front of Kiali.

**Comment by Leo7654** [comment]
On my Mac, echo is working weird. 
```
echo -n aaa
-n aaa
```

**Comment by jdomag** [comment]
Doesn't work for me as well with istio 1.5.2 - password and username has been mounted as below. Does extra empty line or "%" is relevant? 
Secret has been generated using commands for ZSH provided here:
https://istio.io/docs/tasks/observability/kiali/

![image](https://user-images.githubusercontent.com/28705876/80378199-99830e00-889c-11ea-95b7-10114c030f35.png)

**Comment by jmazzitelli** [comment]
@jdomag that secret doesn't look correct. Make sure when you build the secret no newlines or any special characters are on the values. When I do those commands:

```
$ kubectl exec -it -n istio-system kiali-78495d956b-6ztd4 -- cat /kiali-secret/username
admin$ kubectl exec -it -n istio-system kiali-78495d956b-6ztd4 -- cat /kiali-secret/passphrase
admin
```
Notice no newlines or special characters. So it sounds like the secret is not generated properly

Here's what my YAML looks like:

```
$ kubectl get secret kiali -o yaml
apiVersion: v1
data:
  passphrase: YWRtaW4=
  username: YWRtaW4=
kind: Secret
metadata:
  creationTimestamp: "2020-04-27T14:04:57Z"
  labels:
    app: kiali
  name: kiali
  namespace: istio-system
  resourceVersion: "5087902"
  selfLink: /api/v1/namespaces/istio-system/secrets/kiali
  uid: 0d2dc624-1b6e-4ca5-b6de-4a9e6dfbcbc4
type: Opaque
```

**Comment by jdomag** [comment]
@jmazzitelli 
I think this is just how ZSH parsing the output as kubectl get secret return the same result. In fact i've just copied your definition of Secret and restarted Kiali pod - still doesnt work :/

```
kubectl get secret kiali -o yaml                                                     
apiVersion: v1
data:
  passphrase: YWRtaW4=
  username: YWRtaW4=
kind: Secret
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"passphrase":"YWRtaW4=","username":"YWRtaW4="},"kind":"Secret","metadata":{"annotations":{},"labels":{"app":"kiali"},"name":"kiali","namespace":"istio-system"},"type":"Opaque"}
  creationTimestamp: "2020-04-27T14:12:54Z"
  labels:
    app: kiali
  name: kiali
  namespace: istio-system
  resourceVersion: "1207714"
  selfLink: /api/v1/namespaces/istio-system/secrets/kiali
  uid: b1345988-8e41-4d9d-a589-7ef1c7480e2b

```

**Comment by israel-hdez** [comment]
@jdomag How are you exposing/accession Kiali?

**Comment by jdomag** [comment]
i run kubectl proxy and access it via:
`http://localhost:8001/api/v1/namespaces/istio-system/services/kiali:20001/proxy/kiali/`

**Comment by israel-hdez** [comment]
Please, do not use `kubectl proxy` because it has a conflict with Kiali Authentication headers (see explanation here: https://github.com/kiali/kiali/issues/1059#issuecomment-509848506).

You must try another exposure method. See the FAQ for some suggestions: https://kiali.io/faq/general/#how-do-i-access-kiai

**Comment by jdomag** [comment]
@israel-hdez 
Thx, i exposed it via "kubectl port-forward svc/kiali 20001:20001 -n istio-system"  and worked fine.
